File: engine/static/scripting/src/bind.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2024, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "argus/lowlevel/debug.hpp"

#include "argus/core/engine.hpp"

#include "argus/scripting/bind.hpp"
#include "argus/scripting/error.hpp"
#include "argus/scripting/types.hpp"

#include "internal/scripting/module_scripting.hpp"
#include "argus/scripting/util.hpp"

#include <set>
#include <string>
#include <unordered_set>

namespace argus {
    Result<BoundTypeDef, BindingError> create_type_def(const std::string &name, size_t size, const std::string &type_id,
            bool is_refable,
            CopyCtorProxy copy_ctor,
            MoveCtorProxy move_ctor,
            DtorProxy dtor) {
        if (size == 0) {
            crash("Bound types cannot be zero-sized");
        }

        BoundTypeDef def {
                name,
                size,
                type_id,
                is_refable,
                copy_ctor,
                move_ctor,
                dtor,
                {},
                {},
                {},
                {}
        };
        return ok<BoundTypeDef, BindingError>(def);
    }

    Result<BoundEnumDef, BindingError> create_enum_def(const std::string &name, size_t width,
            const std::string &type_id) {
        return ok<BoundEnumDef, BindingError>(BoundEnumDef { name, width, type_id, {}, {} });
    }

    Result<void, BindingError> add_enum_value(BoundEnumDef &def, const std::string &name, int64_t value) {
        if (def.values.find(name) != def.values.cend()) {
            return err<void, BindingError>(BindingErrorType::DuplicateName, def.name + "::" + name,
                    "Enum value with same name is already bound");
        }

        if (def.all_ordinals.find(value) != def.all_ordinals.cend()) {
            return err<void, BindingError>(BindingErrorType::Other, def.name + "::" + name,
                    "Enum value with same ordinal is already bound");
        }

        def.values.insert({ name, value });
        def.all_ordinals.insert(value);

        return ok<void, BindingError>();
    }

    Result<void, BindingError> add_member_instance_function(BoundTypeDef &type_def, const BoundFunctionDef &fn_def) {
        if (type_def.instance_functions.find(fn_def.name) != type_def.instance_functions.cend()) {
            auto qual_name = get_qualified_function_name(FunctionType::MemberInstance, type_def.name, fn_def.name);
            return err<void, BindingError>(BindingErrorType::DuplicateName, qual_name,
                    "Instance function with same name is already bound");
        }

        type_def.instance_functions.insert({ fn_def.name, fn_def });

        return ok<void, BindingError>();
    }

    Result<void, BindingError> add_member_static_function(BoundTypeDef &type_def, const BoundFunctionDef &fn_def) {
        if (type_def.static_functions.find(fn_def.name) != type_def.static_functions.cend()) {
            auto qual_name = get_qualified_function_name(FunctionType::MemberStatic, type_def.name, fn_def.name);
            return err<void, BindingError>(BindingErrorType::DuplicateName, qual_name,
                    "Static function with same name is already bound");
        }

        type_def.static_functions.insert({ fn_def.name, fn_def });

        return ok<void, BindingError>();
    }

    Result<void, BindingError> add_extension_function(BoundTypeDef &type_def, const BoundFunctionDef &fn_def) {
        if (fn_def.params.empty()
                || !(fn_def.params[0].type == IntegralType::Struct || fn_def.params[0].type == IntegralType::Pointer)
                || fn_def.params[0].type_id != type_def.type_id) {
            auto qual_name = get_qualified_function_name(FunctionType::Extension, type_def.name, fn_def.name);
            return err<void, BindingError>(BindingErrorType::InvalidDefinition, qual_name,
                    "First parameter of extension function must match extended type");
        }

        if (type_def.extension_functions.find(fn_def.name) != type_def.extension_functions.cend()) {
            auto qual_name = get_qualified_function_name(FunctionType::Extension, type_def.name, fn_def.name);
            return err<void, BindingError>(BindingErrorType::DuplicateName, qual_name,
                    "Extension function with same name is already bound");
        }

        if (type_def.instance_functions.find(fn_def.name) != type_def.instance_functions.cend()) {
            auto qual_name = get_qualified_function_name(FunctionType::Extension, type_def.name, fn_def.name);
            return err<void, BindingError>(BindingErrorType::ConflictingName, qual_name,
                    "Instance function with same name is already bound");
        }

        type_def.extension_functions.insert({ fn_def.name, fn_def });

        return ok<void, BindingError>();
    }

    Result<void, BindingError> add_member_field(BoundTypeDef &type_def, const BoundFieldDef &field_def) {
        if (field_def.m_type.type == IntegralType::Callback) {
            auto qual_name = get_qualified_field_name(type_def.name, field_def.m_name);
            return err<void, BindingError>(BindingErrorType::InvalidDefinition, qual_name,
                    "Callback-typed fields are not supported");
        }

        if (type_def.fields.find(field_def.m_name) != type_def.fields.cend()) {
            auto qual_name = get_qualified_field_name(type_def.name, field_def.m_name);
            return err<void, BindingError>(BindingErrorType::DuplicateName, qual_name,
                    "Field with same name is already bound for type");
        }

        type_def.fields.insert({ field_def.m_name, field_def });

        return ok<void, BindingError>();
    }
}

File: engine/modules/scripting/src/bind.rs
use std::collections::{HashMap, HashSet};
use std::fmt;
use std::fmt::{Debug, Formatter};
use argus_scripting_bind::*;
use crate::error::{BindingError, BindingErrorType};
use crate::get_qualified_field_name;
use crate::util::get_qualified_function_name;

pub type CloneProxy = fn(*mut (), *const ());
pub type DropProxy = fn(*mut ());

#[derive(Clone)]
pub struct BoundTypeDef {
    pub name: String,
    pub size: usize,
    pub type_id: String,
    // whether references to the type can be passed to scripts
    // (i.e. whether the type derives from AutoCleanupable)
    pub is_refable: bool,
    // the copy and move ctors and dtor are only used for struct value and callback types
    pub cloner: Option<CloneProxy>,
    pub dropper: Option<DropProxy>,
    pub instance_functions: HashMap<String, BoundFunctionDef>,
    pub extension_functions: HashMap<String, BoundFunctionDef>,
    pub static_functions: HashMap<String, BoundFunctionDef>,
    pub fields: HashMap<String, BoundFieldDef>,
}

impl Debug for BoundTypeDef {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), fmt::Error> {
        #[allow(dead_code)]
        #[derive(Debug)]
        struct BoundTypeDef<'a> {
            pub name: &'a String,
            pub size: &'a usize,
            pub type_id: &'a String,
            pub is_refable: &'a bool,
            pub instance_functions: &'a HashMap<String, BoundFunctionDef>,
            pub extension_functions: &'a HashMap<String, BoundFunctionDef>,
            pub static_functions: &'a HashMap<String, BoundFunctionDef>,
            pub fields: &'a HashMap<String, BoundFieldDef>,
        }

        let Self {
            name,
            size,
            type_id,
            is_refable,
            cloner: _,
            dropper: _,
            instance_functions,
            extension_functions,
            static_functions,
            fields,
        } = self;

        Debug::fmt(&BoundTypeDef {
            name,
            size,
            type_id,
            is_refable,
            instance_functions,
            extension_functions,
            static_functions,
            fields,
        }, f)
    }
}

impl BoundTypeDef {
    pub fn new(
        name: impl Into<String>,
        size: usize,
        type_id: impl Into<String>,
        is_refable: bool,
        cloner: Option<CloneProxy>,
        dropper: Option<DropProxy>,
    ) -> Self {
        assert_ne!(size, 0, "Bound types cannot be zero-sized");

        BoundTypeDef {
            name: name.into(),
            size,
            type_id: type_id.into(),
            is_refable,
            cloner,
            dropper,
            instance_functions: HashMap::new(),
            extension_functions: HashMap::new(),
            static_functions: HashMap::new(),
            fields: HashMap::new(),
        }
    }

    pub fn add_field(&mut self, field_def: BoundFieldDef)
        -> Result<(), BindingError> {
        if self.fields.contains_key(&field_def.name) {
            let qual_name = get_qualified_field_name(
                &field_def.name,
                &self.name,
            );
            return Err(BindingError {
                ty: BindingErrorType::DuplicateName,
                bound_name: qual_name,
                msg: "Field with same name is already bound".to_string(),
            });
        }

        self.fields.insert(field_def.name.clone(), field_def);

        Ok(())
    }
    
    pub fn add_instance_function(&mut self, fn_def: BoundFunctionDef)
        -> Result<(), BindingError> {
        if self.instance_functions.contains_key(&fn_def.name) {
            let qual_name = get_qualified_function_name(
                FunctionType::MemberInstance,
                &fn_def.name,
                Some(&self.name)
            );
            return Err(BindingError {
                ty: BindingErrorType::DuplicateName,
                bound_name: qual_name,
                msg: "Instance function with same name is already bound".to_string(),
            });
        }

        self.instance_functions.insert(fn_def.name.clone(), fn_def);

        Ok(())
    }

    pub fn add_static_function(&mut self, fn_def: BoundFunctionDef)
        -> Result<(), BindingError> {
        if self.static_functions.contains_key(&fn_def.name) {
            let qual_name = get_qualified_function_name(
                FunctionType::MemberStatic,
                &fn_def.name,
                Some(&self.name)
            );
            return Err(BindingError {
                ty: BindingErrorType::DuplicateName,
                bound_name: qual_name,
                msg: "Static function with same name is already bound".to_string(),
            });
        }

        self.static_functions.insert(fn_def.name.clone(), fn_def);

        Ok(())
    }

    pub fn add_extension_function(&mut self, fn_def: BoundFunctionDef)
        -> Result<(), BindingError> {
        if fn_def.param_types.is_empty() ||
            !(
                fn_def.param_types[0].ty == IntegralType::Object ||
                    fn_def.param_types[0].ty == IntegralType::Reference ||
                    fn_def.param_types[0].type_id
                        .as_ref()
                        .map(|id| id != &self.type_id)
                        .unwrap_or(false)
            ) {
            let qual_name = get_qualified_function_name(
                FunctionType::Extension,
                &fn_def.name,
                Some(&self.name)
            );
            return Err(BindingError {
                ty: BindingErrorType::InvalidDefinition,
                bound_name: qual_name,
                msg: "First parameter of extension function must match extended type".to_string(),
            });
        }

        if !self.extension_functions.contains_key(&fn_def.name) ||
            self.instance_functions.contains_key(&fn_def.name) {
            let qual_name = get_qualified_function_name(
                FunctionType::MemberStatic,
                &fn_def.name,
                Some(&self.name)
            );
            return Err(BindingError {
                ty: BindingErrorType::DuplicateName,
                bound_name: qual_name,
                msg: "Extension or instance function with same name is already bound".to_string(),
            });
        }

        self.extension_functions.insert(fn_def.name.clone(), fn_def);

        Ok(())
    }
}

#[derive(Clone, Debug)]
pub struct BoundFunctionDef {
    pub name: String,
    pub ty: FunctionType,
    pub is_const: bool,
    pub param_types: Vec<ObjectType>,
    pub return_type: ObjectType,
    pub proxy: ProxiedNativeFunction,
}

impl BoundFunctionDef {
    pub fn new(
        name: impl Into<String>,
        ty: FunctionType,
        is_const: bool,
        param_types: Vec<ObjectType>,
        return_type: ObjectType,
        proxy: ProxiedNativeFunction,
    ) -> Self {
        Self {
            name: name.into(),
            ty,
            is_const,
            param_types,
            return_type,
            proxy,
        }
    }
}

#[derive(Clone, Debug)]
pub struct BoundFieldDef {
    pub name: String,
    pub ty: ObjectType,
    //TODO: we probably only need one function for getting the value
    pub access_proxy: fn(&WrappedObject, &ObjectType) -> WrappedObject,
    pub assign_proxy: Option<fn(&mut WrappedObject, &WrappedObject)>,
}

impl BoundFieldDef {
    pub fn new(
        name: impl Into<String>,
        ty: ObjectType,
        access_proxy: fn(&WrappedObject, &ObjectType) -> WrappedObject,
        assign_proxy: Option<fn(&mut WrappedObject, &WrappedObject)>,
    ) -> Self {
        Self {
            name: name.into(),
            ty,
            access_proxy,
            assign_proxy,
        }
    }
    
    pub fn get_value(&self, instance: &mut WrappedObject) -> WrappedObject {
        (self.access_proxy)(instance, &instance.ty)
    }

    pub fn set_value(&self, instance: &mut WrappedObject, value: &WrappedObject) {
        self.assign_proxy.unwrap()(instance, value);
    }
}

#[derive(Clone, Debug)]
pub struct BoundEnumDef {
    pub name: String,
    pub width: usize,
    pub type_id: String,
    pub values: HashMap<String, i64>,
    pub all_ordinals: HashSet<i64>,
}

impl BoundEnumDef {
    pub fn new(
        name: impl Into<String>,
        width: usize,
        type_id: impl Into<String>
    ) -> Self {
        BoundEnumDef {
            name: name.into(),
            width,
            type_id: type_id.into(),
            values: HashMap::new(),
            all_ordinals: HashSet::new(),
        }
    }

    pub fn add_value(&mut self, name: impl AsRef<str>, value: i64) -> Result<(), BindingError> {
        if self.values.contains_key(name.as_ref()) {
            return Err(BindingError {
                ty: BindingErrorType::DuplicateName,
                bound_name: format!("{}::{}", self.name, name.as_ref()),
                msg: "Enum value with same name is already bound".to_string(),
            });
        }

        if self.all_ordinals.contains(&value) {
            return Err(BindingError {
                ty: BindingErrorType::Other,
                bound_name: format!("{}::{}", self.name, name.as_ref()),
                msg: "Enum value with same ordinal is already bound".to_string()
            });
        }

        self.values.insert(name.as_ref().to_string(), value);
        self.all_ordinals.insert(value);

        Ok(())
    }
}
